tasks.py

```diff
@@ -143,39 +185,50 @@ def logAsRetried(message, *args, **kwargs):
     logging.log(level, message, *args, **kwargs)
 
 
-def _defer(queues, func, funcArgs, funcKwargs, countdown=None, eta=None, 
-           taskName=None, target=None, transactional=False, retry_options=None):
+def _defer(queues, func, funcArgs, funcKwargs, countdown=None, eta=None,
+           taskName=None, target=None, transactional=False, retry_options=None, parent=None):
     """
     Our own implementation of deferred.defer.
-    
+
     This allows:
     - using webapp2 as deferred handler and applying our middlewares
     - using task asynchronous API
     - using cPickle instead of pickle
     - logging headers at DEBUG level instead of INFO
     """
     payload = _serialize(func, funcArgs, funcKwargs)
-    
+
     queueName = random.choice(queues)
-    
+
     # We track which function is called so that it appears clearly in the App admin dash-board.
     # Note: if it's a class method, we only track the method name and not the class name.
     url = "/_cb/deferred/%s/%s" % (getattr(func, '__module__', ''), getattr(func, '__name__', ''))
-    
+
     headers = {"Content-Type": "application/octet-stream"}
-    
-    task = taskqueue.Task(payload=payload, target=target, url=url, headers=headers,
-                          countdown=countdown, eta=eta, name=taskName, retry_options=retry_options)
-    
+
+    try:
+        task = taskqueue.Task(payload=payload, target=target, url=url, headers=headers,
+                              countdown=countdown, eta=eta, name=taskName, retry_options=retry_options)
+    except taskqueue.TaskTooLargeError:
+        logging.info('Task Too Large. Storing payload in the data store')
+        if parent:
+            # In case if data store entity needs to be stored under any specific parent
+            key = _DeferredTaskEntity(data=payload, parent=parent).put()
+        else:
+            key = _DeferredTaskEntity(data=payload).put()
```

Comment:
the _defer function returns a Future, and must include all operations inside. Otherwise it's no longer really an async operation and could even lead to stackoverflow errors when you enqueue a lot of tasks.

To make this change, we need to make `_defer` a tasklet,  yield  the `put_async` operation and yield the `task.add_async` operation.

Reply:
also, no need to make a special case for parent, you can just always pass it to _DeferredTaskEntity. If it's None it'll just do nothing.